ci/bin/format.dart

```diff
@@ -0,0 +1,795 @@
+
+// Checks and fixes format on files with changes.
+//
+// Usage: dart format.dart [--fix] [--all]
+//   --fix will fix all of the formatting problems found.
+//   --all will find/fix formatting problems for all files, not just changed
+//         files.
+
+import 'dart:io';
+
+import 'package:args/args.dart';
+import 'package:meta/meta.dart';
+import 'package:path/path.dart' as path;
+import 'package:process/process.dart';
+import 'package:process_runner/process_runner.dart';
+
+class FormattingException implements Exception {
+  FormattingException(this.message, [this.result]);
+
+  final String message;
+  final ProcessResult /*?*/ result;
+
+  int get exitCode => result?.exitCode ?? -1;
+
+  @override
+  String toString() {
+    String output = runtimeType.toString();
+    output += ': $message';
+    final String stderr = result?.stderr as String;
+    if (stderr.isNotEmpty) {
+      output += ':\n$stderr';
+    }
+    return output;
+  }
+}
+
+enum MessageType {
+  message,
+  error,
+  warning,
+}
+
+enum FormatCheck {
+  clang,
+  java,
+  whitespace,
+  gn,
+}
+
+FormatCheck nameToFormatCheck(String name) {
+  switch (name) {
+    case 'clang':
+      return FormatCheck.clang;
+    case 'java':
+      return FormatCheck.java;
+    case 'whitespace':
+      return FormatCheck.whitespace;
+    case 'gn':
+      return FormatCheck.gn;
+  }
+  assert(false, 'Unknown FormatCheck type $name');
+  return null;
+}
+
+String formatCheckToName(FormatCheck check) {
+  switch (check) {
+    case FormatCheck.clang:
+      return 'C++/ObjC';
+    case FormatCheck.java:
+      return 'Java';
+    case FormatCheck.whitespace:
+      return 'Trailing whitespace';
+    case FormatCheck.gn:
+      return 'GN';
+  }
+  assert(false, 'Unhandled FormatCheck type $check');
+  return null;
+}
+
+List<String> formatCheckNames() {
+  return FormatCheck.values
+      .map<String>((FormatCheck check) => check.toString().replaceFirst('$FormatCheck.', ''))
+      .toList();
+}
+
+typedef MessageCallback = Function(String message, {MessageType type});
+
+/// Base class for format checkers.
+///
+/// Provides services that all format checkers need.
+abstract class FormatChecker {
+  FormatChecker({
+    ProcessManager /*?*/ processManager,
+    @required this.baseGitRef,
+    @required this.repoDir,
+    @required this.srcDir,
+    this.allFiles = false,
+    this.messageCallback,
+  }) : _processRunner = ProcessRunner(
+          defaultWorkingDirectory: repoDir,
+          processManager: processManager ?? const LocalProcessManager(),
+        ) {
+    /*late*/ String clangOs;
+    if (Platform.isLinux) {
+      clangOs = 'linux-x64';
+    } else if (Platform.isMacOS) {
+      clangOs = 'mac-x64';
+    } else {
+      throw FormattingException(
+          "Unknown operating system: don't know how to run clang-format here.");
+    }
+    clangFormat = File(
+        path.join(srcDir.absolute.path, 'buildtools', clangOs, 'clang', 'bin', 'clang-format'));
```

Comment:
nit: add a trailing comma for nicer formatting

Reply:
Done.